// marko/dist/runtime/helpers/tags-compat/runtime-dom.js
"use strict";const { _k_ } = require("@internal/components-util");
const {
  S_
} = require("../../components/ComponentsContext");
const defineComponent = require("../../components/defineComponent");
const { r: registerComponent } = require("../../components/registry");
const createRenderer = require("../../components/renderer");
const defaultCreateOut = require("../../createOut");
const morphdom = require("../../vdom/morphdom");
const { _j_ } = require("../../vdom/morphdom/fragment");
const dynamicTag = require("../dynamic-tag");

exports.p = function (domCompat) {
  dynamicTag.bl_ = function tagsToVdom(
  renderer,
  renderBody,
  args)
  {
    const tagsRenderer = domCompat.resolveRenderer(renderer || renderBody);

    if (tagsRenderer) {
      return (input, out) =>
      TagsCompat({ i: args ? args : input, r: tagsRenderer }, out);
    }

    return renderer;
  };

  const TagsCompatId = "tags-compat";
  const TagsCompat = createRenderer(
    function (_, out, componentDef, component) {
      const isHydrate =
      S_(out).p_._Y_;
      const input = Array.isArray(_.i) ? _.i : [_.i];
      const tagsRenderer = domCompat.resolveRenderer(_.r);
      const newNode = domCompat.render(
        isHydrate,
        out,
        component,
        tagsRenderer,
        input
      );

      out.bf(out.a__, component, !newNode);
      if (newNode) {
        out.node({ bm_: () => newNode });
      }
      out.ef();
    },
    // eslint-disable-next-line no-constant-condition





    {
      t: TagsCompatId
    },
    {}
  );

  registerComponent(TagsCompatId, () => ({
    _: TagsCompat,
    Component: defineComponent(
      {
        onMount: domCompat.runComponentEffects,
        onUpdate: domCompat.runComponentEffects
      },
      TagsCompat
    )
  }));

  // (
  //   nodeAccessor: Accessor,
  //   dynamicTagAttrs?: IntersectionSignal,
  //   intersection?: IntersectionSignal,
  //   valueWithIntersection?: ValueSignal
  // )

  const rendererCache = new WeakMap();

  domCompat.patchConditionals((conditional) => (...args) => {
    const signal = conditional(...args);
    const hasAttrs = args.length > 1;
    return (scope, renderer, clean) => {
      return signal(scope, create5to6Renderer(renderer, hasAttrs), clean);
    };
  });

  function create5to6Renderer(renderer, hasAttrs) {
    let newRenderer = renderer;
    if (renderer && typeof renderer !== "string") {
      const rendererFromAnywhere =
      renderer._ ||
      renderer.render ||
      renderer.renderer && renderer.renderer.renderer ||
      renderer.renderer;

      if (!domCompat.isRenderer(rendererFromAnywhere || renderer)) {
        newRenderer = rendererCache.get(renderer);
        if (!newRenderer) {
          const { Component } = renderer;
          if (Component) {
            const setCustomEvents = Component.prototype.aC_;
            Component.prototype.aC_ = function (
            customEvents,
            scopeId)
            {
              for (const customEvent of customEvents) {
                customEvent[1] = domCompat.resolveRenderer(customEvent[1]);
              }

              setCustomEvents.call(this, customEvents, scopeId);
            };
          }
          newRenderer = domCompat.createRenderer(
            (scope) => {
              if (!hasAttrs) {
                renderAndMorph(scope, rendererFromAnywhere, renderer, {});
              }
            },
            () => {
              const realFragment = document.createDocumentFragment();
              _j_(null, null, realFragment);
              return realFragment;
            },
            (scope, input, clean) => {
              if (clean) return;
              renderAndMorph(scope, rendererFromAnywhere, renderer, input);
            }
          );
          rendererCache.set(renderer, newRenderer);
        }
      }
    }
    return newRenderer;
  }

  domCompat.register("@marko/tags-compat-5-to-6", create5to6Renderer);

  function renderAndMorph(scope, renderer, renderBody, input) {
    const out = defaultCreateOut();
    let host = domCompat.getStartNode(scope);
    let rootNode = host.fragment;
    if (!rootNode) {
      const component = scope.marko5Component = _k_[scope.m5c];
      rootNode = component._C_;
      host = rootNode.startNode;
      domCompat.setScopeNodes(host, rootNode.endNode);
    }
    const existingComponent = scope.marko5Component;
    const componentsContext = S_(out);
    const globalComponentsContext = componentsContext.p_;
    let customEvents;
    globalComponentsContext.ay_ = existingComponent;
    out.sync();
    if (renderer) {
      const [rawInput] = input;
      const normalizedInput = {};

      for (const key in rawInput) {
        let value = rawInput[key];
        if (key.startsWith("on")) {
          const c = key[2];
          customEvents = customEvents || {};
          customEvents[(c === "-" ? "" : c.toLowerCase()) + key.slice(3)] = [
          value];

        } else {
          normalizedInput[key] = rawInput[key];
        }
      }

      renderer(normalizedInput, out);
    } else {
      RenderBodyComponent({ renderBody, args: input }, out);
    }

    domCompat.queueEffect(scope, () => {
      const targetNode = out.ag_().az_;
      morphdom(rootNode, targetNode, host, componentsContext);
      const componentDefs = componentsContext.ae_(
        getRootNode(host)
      );
      const component = componentDefs[0].s_;
      component._C_ = rootNode;
      component.O_ = input[0];
      component.V_ = customEvents;
      scope.marko5Component = component;
    });
  }

  function getRootNode(el) {
    var cur = el;
    while (cur.parentNode) cur = cur.parentNode;
    return cur;
  }

  const RenderBodyComponentId = "renderbody-renderer";
  const RenderBodyComponent = createRenderer(
    function (input, out, _componentDef) {
      dynamicTag(
        out,
        input.renderBody,
        null,
        null,
        input.args,
        null,
        _componentDef,
        "0"
      );
    },
    // eslint-disable-next-line no-constant-condition






    {
      t: RenderBodyComponentId,
      i: true
    },
    {}
  );

  registerComponent(RenderBodyComponentId, () => ({
    _: RenderBodyComponent,
    Component: defineComponent({}, RenderBodyComponent)
  }));
};
